### ISSUE TITLE
Up-streaming Map-overlap Functions

### ISSUE THREAD

**Issue by CSSFrancis** [hyperspy/hyperspy#3128]
#### Describe the functionality you would like to see.
There are a couple of functions using map-overlap in packages that extend hyperspy and I was thinking of up streaming their implementation to hyperspy.  I figured we could give hyperspy the right of first refusal before implementing downstream.

Adding support for the [scipy.ndimage](https://docs.scipy.org/doc/scipy/reference/ndimage.html) functions in particular would be quite beneficial.  These are already implemented lazily in [dask-image](https://image.dask.org/en/latest/) but I would also like to consider a slightly different option which allows users to have the option to reload data from memory rather than holding memory in cache.  This is helpful for when you have many cores but are limited by the amount of memory per core as map_overlap functions are notorious for loading the entire dataset into RAM.  

Some ideas of functions that might be universally useful:
- Gaussian Filtering in all dimensions
- Difference of Gaussian in n-dimensions to reduce noise 
- Laplacian of Gaussian to find edges in n-dimensions
- Maximum/ Minimum filter to find peaks in n-dimensions
- Filter with a custom filter

#### Describe the context
I propose we add a generic function `map_overlap` function that applies some function to some or all of the axes. 

```
def _map_overlap(self, func, overlaps, boundary, reload_tasks=False, **kwargs):
    data_overlapped = da.overlap.overlap(self.data, 
                                                                 depth=overlaps,
                                                                 boundary=boundary)
    if reload_tasks:
        # Clone original overlapped dask blocks to reload data from disk and stop spilling to disk
        data_clones = da.concatenate(
            [clone(b, omit=data_overlapped) for b in data_overlapped.blocks]
        )
    else:
        data_clones = data_overlapped

    mapped = data_clones.map_blocks(func, **kwargs)
    return self.deep_copy_with_new_data(mapped)
```

#### Additional information

The `reload_task` option will only reload tasks along one of the chunked dimension.  This means that for a 4 dimensional dataset it will load chunk a, b,e and f then operate on that chunk.  Then it will load c and g while holding a,b,e and f in memory.  Then it will drop a and e from memory while loading d, h into memory. 

This means that each core will only ever have 9 chunks in memory at a time. This differs from normal `dask` operation in that the total RAM won't spike when using an overlapping function.  Normally all of the chunks are loaded into ram during an overlapping computation, especially when using multiple cores because for each chunk it sees that it is needed for some other operation and as a result holds that chunk in RAM until it is needed. 

| a | b | c | d |
+--+--+--+--+
| e | f | g | h |
+--+--+--+--+
| i | k | k | l |

The argument against this workflow is:  

1. It gets really slow if some operation is applied to the data besides just loading that data.  That operation would have to be repeated 3x as the entire task tree for the block is cloned.
2. It can potentially be slower if you can load the entire dataset into RAM as you don't have the extra overhead for reloading the data from disk
3. Ideally you would be able to spill loaded chunks to disk in such a way as to make this workflow unnecessary.  In practice spilling to disk has some unintended consequences and causes operations to crash. 

Another possibility is to force only one axis to be chunked.  This reduces the complexity but the size of chunks can get very large and cause additional issues.

**Comment by ericpre** [comment]
Yes, it makes sense to add a `map_overlap` method, similar to the `map` method.

Regarding the use of `dask-image`, I am wondering if we could have something similar to the `_map_all` functions by passing `dask-image` functions:

```python
s = hs.signals.Signal2D(array).as_lazy()
s.map(dask_image.ndfilters.gaussian_filter, sigma=3)
```

I am thinking about this because as noted in https://github.com/hyperspy/hyperspy/pull/3124 `scipy` recently added the axes argument to ndfilters, which means that we can use `_map_all` directly, i.e. apply the function directly to the data without iterating on the navigation dimension. For non-lazy signals, no changes is required in hyperspy to benefit from these improvement. 😃
I expect that the same approach (use `_map_all`) would be possible by using `dask_image.ndfilters.gaussian_filter` in place of `ndfilters.gaussian_filter` with little changes to the `map` method - at the moment it is only for non-lazy signal but it would be possible to add cases that check that the function is from `dask-image` and check that this is suitable. It don't know if it would be necessary to add support of the `axes` argument to `dask_image.ndfilters.gaussian_filter` to get this to work but maybe this would be a good way to use  `dask-image`.

It may be sensible to have both approaches, as one is generic and the other allow to use the `dask_image` library which can be used to work lazily with cupy array out of the box.
